source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "source/extensions/config_subscription/grpc/grpc_stream.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * This class arbitrates between two config providers of the same GrpcMux -
+ * the primary and the failover. Envoy always prefers fetching config from the
+ * primary source, but if not available, will fetch the config from the failover
+ * source until the primary is again available.
+ *
+ * This class owns the state for the GrpcMux primary and failover streams, and
+ * proxies the gRPC stream functionality to either the primary or the failover config sources.
+ * The failover source is optional and will only be used if passed upon initialization.
+ *
+ * The primary config source is always preferred over the failover source. If the primary
+ * is available, the stream to the failover source will terminated.
+ *
+ * Failover is supported in both
+ * SotW (RequestType = envoy::service::discovery::v3::DiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
+ *
+ * The use of this class will be as follows: the GrpcMux object will own an instance of
+ * the GrpcMuxFailoverManager. The GrpcMuxFailoverManager will own 2 GrpcStreams, primary
+ * and secondary. Each of the primary and secondary streams will invoke GrpcStreamCallbacks
+ * on their corresponding objects (also owned by the GrpcMuxFailoverProxy). These invocations
+ * will be followed by the GrpcMuxFailoverProxy calling the GrpcStreamCallbacks on the GrpcMux
+ * object that initialized it.
+ *
+ * Note: this class is WIP.
+ */
+template <class RequestType, class ResponseType> class GrpcMuxFailover {
+public:
+  // A GrpcStream creator function that receives the stream callbacks and returns a
```

Comment:
Should this be (named) a factory class / pattern?

Reply:
Not sure if I fully grok the suggestion.
The underlying issue is tight coupling between different classes (somewhat similar to the tight coupling currently between GrpcMux and GrpcStream, and I'm guessing that is why there was no mock for GrpcStream up to this point).
In our case the GrpcMuxFailover creates and owns the grpc-stream-callbacks handling object (e.g., `PrimaryGrpcStreamCallbacks` for the primary) and the GrpcStream object that is initialized with the callbacks, but to create the PrimaryGrpcStreamCallbacks object, it needs a pointer to the GrpcMuxFailover.

Let me know if you want to discuss this further IRL.